### DISCUSSION TITLE
After use clear i dont get empty array in watch

### DISCUSSION THREAD

**Discussion by Enhed** [flutterdata/flutter_data#192] [Q&A] [answered]
Clear realy deletes all items in repository, but i dont get changes in `watchAll`, only if i close page and entire again - list will be empty. What im doing wrong?

**Comment by frank06** [comment, marked as answer]
Can you update to 1.5.3 and see if that fixes it?

**Reply by frank06** [reply]
https://github.com/flutterdata/flutter_data/commit/574ecec523b4de4fae43b59dfc563482c9d33eac

**Reply by Enhed** [reply]
on `1.5.3` same behavior

**Reply by frank06** [reply]
The test would indicate it has to work.

https://github.com/flutterdata/flutter_data/blob/574ecec523b4de4fae43b59dfc563482c9d33eac/test/repository/remote_adapter_watch_test.dart#L444-L448

Could you share some code to investigate this further?

**Reply by frank06** [reply]
Otherwise you could try looping through your models and deleting one by one

**Reply by Enhed** [reply]
@frank06 I think I will stop using `watchers` anyway because even on 1-2 `items` in `ListView` I get severe lags over time and do not know how to fix it

**Reply by Enhed** [reply]
```dart
import 'package:flutter_data/flutter_data.dart';
import 'package:json_annotation/json_annotation.dart';

part 'log.g.dart';

@JsonSerializable()
@DataRepository([], remote: false)
class Log extends DataModel<Log> {
  @override
  final DateTime id;
  final String message;

  Log(this.id, this.message);
}
```

```dart
import 'package:flutter/material.dart';
import 'package:flutter_data/flutter_data.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:hooks_riverpod/hooks_riverpod.dart';
import 'package:sync_test/main.data.dart';

import 'models/log.dart';

void main() {
  runApp(ProviderScope(
    overrides: [configureRepositoryLocalStorage()],
    child: App(),
  ));
}

class App extends HookConsumerWidget {
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text('Logs')
        ),

        body: Center(
          child: ref.watch(repositoryInitializerProvider).when(
            error: (error, _) => Text(error.toString()),

            loading: () => const CircularProgressIndicator(),

            data: (_) {
              final state = ref.logs.watchAll();

              if(state.isLoading) {
                return const CircularProgressIndicator();
              }

              final logs = state.model ?? []; // empty if no data

              return Column(
                children: [
                  Text('Have x${logs.length} logs'),

                  ElevatedButton(onPressed: ref.logs.clear, child: Text('clear')),

                  Expanded(
                    child: ListView(
                      children: [
                        for(final log in logs) ListTile(
                          title: Text(log.message),
                          subtitle: Text(log.id.toIso8601String()),
                        )
                      ],
                    ),
                  )
                ],
              );
            },
          ),
        ),

        floatingActionButton: FloatingActionButton(
          onPressed: () => Log(DateTime.now(), 'message').save(),
        ),
      ),
      debugShowCheckedModeBanner: false,
    );
  }
}
```

this is not clearing list if tap on action in clear button

**Reply by frank06** [reply]
You're right, your code is correct. The feature had a bug, which is now fixed in `1.5.4`

This library is being used in production with tons of items (not 1 or 2) in `ListView`s without issues, and has had many performance improvements over time. If you're able to find out what's causing the lag, I would appreciate if you could share it with me

**Reply by Enhed** [reply]
@frank06 examples on same code:


https://user-images.githubusercontent.com/36981278/209415404-e03c9d81-158e-4591-814c-fcdd7933f009.mp4

after wait a bit:


https://user-images.githubusercontent.com/36981278/209415414-0a9b2df8-a85b-492d-ae7e-6945b68bd726.mp4

Basically the problem is accumulative - the longer `watcher` works, the more laggy it is. On a real project with multiple dependent pages and more data volume it is much worse, lags almost immediately

**Reply by frank06** [reply]
If you replaced the watcher with a regular `ListView` with 100+ items does that work better? On iphone or iphone simulator you get the same?

Btw I noticed that you didn't wrap the watcher in a `HookBuilder`, right now every update rebuilds the whole scaffold.

**Reply by Enhed** [reply]
Without `watcher` everything works fine, e.g. through `findAll`. On the `iphone` unfortunately I can't test it. About `HookBuilder` I don't know anything, I'm guided by the documentation, I've never worked with providers before.

**Reply by frank06** [reply]
What could drag down performance? Put breakpoints and see how often the watcher is rebuilding the widget. (I would highly recommend getting familiar with providers because Flutter Data is based on Riverpod.)

There has to be something else at play here, what emulator are you using? Does it have enough resources on your computer? I'm trying your code on a Pixel 2 emulator and it works just fine even after 15+ minutes with 150+ logs. Did you try on device?

**Reply by frank06** [reply]
And did the clear fix work for you?

**Reply by Enhed** [reply]
> And did the clear fix work for you?

yes, all works fine. I will try another emulator, as there are no problems with the resources of the system itself, thanks for the tip

**Reply by Enhed** [reply]
Earlier I ran on Pixel XL API 30 and I thought that because of this, I installed the Pixel 2 API 33 and nothing has changed, after 15 minutes again strong lags. After that I tried Pixel 2 XL API 33 by allocating more system resources to it - problem is reproduced. Do not even know what this may mean, it happinings only when using watcher.

system conf: i5-12500; ram 32GB (free 40%), GPU rtx 3050 without load, ssd m2 500GB 30% space.

p.s. breakpoints showed nothing, no rebuildings before the laggs, even if the list is empty just by calling watchAll it will happen - noticeable by the animation of ListView and app_time_stats increases over time.

I'll try later on a real device.

**Reply by Enhed** [reply]
@frank06 

btw this example is not lagging:  
```dart
import 'package:flutter/material.dart';
import 'package:flutter_data/flutter_data.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:hooks_riverpod/hooks_riverpod.dart';
import 'package:sync_test/main.data.dart';

void main() {
  runApp(ProviderScope(
    overrides: [configureRepositoryLocalStorage()],
    child: App(),
  ));
}

class Log {
  final DateTime id;
  final String message;

  Log({required this.id, required this.message});
}

class LogListState extends StateNotifier<List<Log>> {
  LogListState(): super([]);

  add(String message) => state = [...state, Log(message: message, id: DateTime.now())];
  clear() => state = [];
}

final logListProvider = StateNotifierProvider<LogListState, List<Log>>((ref) => LogListState());

class App extends HookConsumerWidget {
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text('Logs'),
        ),

        body: Center(
          child: ref.watch(repositoryInitializerProvider).when(
            error: (error, _) => Text(error.toString()), //asd

            loading: () => const CircularProgressIndicator(),

            data: (_) {
              final logs = ref.watch(logListProvider);

              return Column(
                children: [
                  Text('Logs count ${logs.length}'),

                  ElevatedButton(onPressed: () => ref.read(logListProvider.notifier).clear(), child: Text('clear')),

                  Expanded(
                    child: ListView(
                      children: [
                        for(final log in logs) ListTile(
                          title: Text(log.message),
                          subtitle: Text(log.id.toIso8601String()),
                        )
                      ],
                    )
                  )
                ],
              );
            },
          ),
        ),

        floatingActionButton: Consumer(builder: (context, ref, child) => FloatingActionButton(
          onPressed: () => ref.read(logListProvider.notifier).add('message'),
        ))
      ),
      debugShowCheckedModeBanner: false,
    );
  }
}
```

**Reply by frank06** [reply]
Thanks. I opened an issue to track this
